### PR TITLE
[batch2] billing v1

### PR DESCRIPTION
**Pull request by jigold**
I added the columns to the batches table for the real-time cost estimate even though that will be v2 to avoid having to reset the database twice. Although, we may need to do that anyways so the fields are properly interpreted.

### CODE REVIEW

File: batch2/create-batch-tables.sql [hunk 1]
```diff
@@ -364,4 +439,68 @@ BEGIN
...
+
+  START TRANSACTION;
+
+  SELECT start_time, end_time
+  INTO cur_start_time, cur_end_time
+  FROM attempts
+  WHERE batch_id = in_batch_id AND job_id = in_job_id AND attempt_id = in_attempt_id;
+
+  IF cur_end_time IS NULL THEN
+    IF cur_start_time IS NOT NULL AND cur_start_time != in_start_time THEN
```
Comment (jigold) [inline batch2/create-batch-tables.sql]: I think I need to get rid of this check because the preemption event may update the database before the other messages are processed.

File: batch2/batch/driver/instance.py [hunk 2]
```diff
@@ -105,6 +105,16 @@ def state(self):
         self._state = 'deleted'
         self.instance_pool.adjust_for_add_instance(self)
 
+    async def mark_jobs_ended(self, timestamp):
+        await self.db.execute_update(
+            '''
+UPDATE attempts
+INNER JOIN jobs ON attempts.batch_id = jobs.batch_id AND attempts.job_id = jobs.job_id
+SET start_time = %s, end_time = %s
```
Comment (cseed) [inline batch2/batch/driver/instance.py]: This seems wrong: you're setting the start to the timestamp.  It looks like this is called on deactivation/preemption.  The cost of the job is from when it was started until when it was preempted.

File: batch2/batch/driver/instance_pool.py [hunk 3]
```diff
@@ -287,20 +287,24 @@ def add_instance(self, instance):
...
             raise
 
-    async def handle_preempt_event(self, instance):
+    async def handle_preempt_event(self, instance, timestamp):
+        await instance.mark_jobs_ended(timestamp)
         await self.call_delete_instance(instance)
 
-    async def handle_delete_done_event(self, instance):
+    async def handle_delete_done_event(self, instance, timestamp):
+        await instance.mark_jobs_ended(timestamp)
```
Comment (cseed) [inline batch2/batch/driver/instance_pool.py]: I think this should only run when the instance is stopped or deleted.  As far as I can tell, Google charges us for the grace period.  A job will continue to run (and maybe finish) while the preemption/deletion is being processed.

File: batch2/batch/batch.py [hunk 4]
```diff
@@ -280,22 +308,25 @@ def job_record_to_dict(record, running_status=None):
...
             'job_id': job_id,
+            'attempt_id': attempt_id,
             'user': record['user'],
             'state': 'error',
             'error': traceback.format_exc(),
             'container_statuses': {k: {} for k in tasks}
         }
-        await mark_job_complete(app, batch_id, job_id, 'Error', status)
+        now = time.time()
+        await mark_job_complete(app, batch_id, job_id, attempt_id, 'Error', status, now, now)
```
Comment (cseed) [inline batch2/batch/batch.py]: I'd set timing to None, None to be consistent with cancellation.  It never started running.

File: batch2/batch/utils.py [hunk 5]
```diff
@@ -9,6 +9,10 @@
 log = logging.getLogger('utils')
 
 
+def format_currency(value):
+    return "${:,.4f}".format(value)
```
Comment (cseed) [inline batch2/batch/utils.py]: You can do this with an f-string: `f'{value:.4f}'`.

File: batch2/batch/worker.py [hunk 6]
```diff
@@ -504,31 +518,53 @@ def id(self):
...
+        self.start_time = start_time
+        asyncio.ensure_future(worker.post_job_timing(self))
+
+    async def mark_cancelled(self, worker, end_time):
+        self.mark_ended(end_time)
+        asyncio.ensure_future(worker.post_job_timing(self))
+
+    def mark_ended(self, end_time):
+        assert self.start_time and not self.end_time
+        self.end_time = end_time
```
Comment (cseed) [inline batch2/batch/worker.py]: It is a little surprising this one doesn't post the job timings, but the other mark_ functions do.

File: batch2/batch/worker.py [hunk 7]
```diff
@@ -714,6 +750,32 @@ def __init__(self):
...
+                    return
+            except asyncio.CancelledError:  # pylint: disable=try-except-raise
+                raise
+            except Exception as e:
+                if isinstance(e, aiohttp.ClientResponseError) and e.status == 404:   # pylint: disable=no-member
+                    raise
+                log.exception(f'failed to post {job} timing, retrying')
+
+            # exponentially back off, up to (expected) max of 30s
+            t = delay * random.uniform(0.7, 1.3)
```
Comment (cseed) [inline batch2/batch/worker.py]: Hmm, is it necessary to infinitely retry job timings, even for errors?  I'm not sure.  I wonder if retry_transient_errors woudl be suitable here.

File: batch2/create-batch-tables.sql [hunk 8]
```diff
@@ -33,6 +33,11 @@ CREATE TABLE IF NOT EXISTS `batches` (
   `n_failed` INT NOT NULL DEFAULT 0,
   `n_cancelled` INT NOT NULL DEFAULT 0,
   `time_created` TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+  `time_completed` TIMESTAMP,
+  `cost` DOUBLE NOT NULL DEFAULT 0,
+  `running_cost` DOUBLE NOT NULL DEFAULT 0,
```
Comment (cseed) [inline batch2/create-batch-tables.sql]: Can you comment of these are for future use?

Or just delete them.  It isn't necessary.  I don't mind blowing away the database for now, and we'll need to stop doing it completely soon anyway.  I'm working on a database migration plan for that.

File: batch2/create-batch-tables.sql [hunk 9]
```diff
@@ -49,19 +54,38 @@ CREATE TABLE IF NOT EXISTS `jobs` (
...
+  FOREIGN KEY (`batch_id`, `job_id`) REFERENCES jobs(batch_id, job_id) ON DELETE CASCADE
+) ENGINE = InnoDB;
+
 CREATE TABLE IF NOT EXISTS `ready_cores` (
   ready_cores_mcpu INT NOT NULL
 ) ENGINE = InnoDB;
 
 INSERT INTO ready_cores (ready_cores_mcpu) VALUES (0);
 
+CREATE TABLE IF NOT EXISTS `gevents` (
```
Comment (cseed) [inline batch2/create-batch-tables.sql]: Sounds like this is a copy of the gevents.  How about gevents_mark?

File: batch2/create-batch-tables.sql [hunk 10]
```diff
@@ -94,6 +118,42 @@ CREATE INDEX batch_attributes_key_value ON `batch_attributes` (`key`, `value`(25
 
 DELIMITER $$
 
+CREATE TRIGGER attempts_before_update BEFORE UPDATE ON attempts
+FOR EACH ROW
+BEGIN
+  IF OLD.end_time IS NOT NULL AND (NEW.end_time IS NULL OR NEW.end_time > OLD.end_time) THEN
```
Comment (cseed) [inline batch2/create-batch-tables.sql]: end_time can't be null and not null.

File: batch2/batch/batch.py [hunk 11]
```diff
@@ -126,6 +149,9 @@ def job_record_to_dict(record, running_status=None):
     if status:
         result['status'] = status
 
+    if record['state'] not in ('Pending', 'Ready'):
+        result['cost'] = format_currency(record['cost'])
```
Comment (cseed) [inline batch2/batch/batch.py]: Same comment as for batches.

File: batch2/batch/batch.py [hunk 12]
```diff
@@ -104,6 +111,22 @@ def batch_record_to_dict(record):
     await notify_batch_job_complete(db, batch_id)
 
 
+async def add_job_timing(app, batch_id, job_id, attempt_id, start_time, end_time):
+    db = app['db']
+
+    id = (batch_id, job_id)
+
+    log.info(f'marking job {id} cancelled')
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch2/batch/batch.py]: I don't quite follow this comment. Can you elaborate?

File: batch2/batch/driver/instance.py (hunk 2 above)
Comment (jigold) [reply batch2/batch/driver/instance.py]: I can set the start_time to Null. I think the cost will be the same. The trigger left-normalizes any updates and then computes the difference between the new time interval and the old time interval. I could be wrong, but I think this trigger is correct and makes the logic much simpler to deal with. Otherwise, theoretically you could get a preemption event, followed by a mark job complete. I wanted the mark job complete to be handled properly in this case.

File: batch2/batch/driver/instance_pool.py (hunk 3 above)
Comment (jigold) [reply batch2/batch/driver/instance_pool.py]: Ok. So it should only be in the event of delete_done and not called_delete or preempted?

Comment (jigold) [reply batch2/batch/driver/instance_pool.py]: Also, see my comment above about the order of events not mattering with the way I set up the trigger.

File: batch2/batch/batch.py (hunk 4 above)
Comment (jigold) [reply batch2/batch/batch.py]: good catch.

File: batch2/batch/worker.py (hunk 7 above)
Comment (jigold) [reply batch2/batch/worker.py]: I'm not sure what the right answer is. For the job_started, it doesn't matter so much, but if a job is cancelled we definitely want that information of how long it had been running.

File: batch2/create-batch-tables.sql (hunk 10 above)
Comment (jigold) [reply batch2/create-batch-tables.sql]: I guess that's true for the end time, but I think we do need this check for the start time.

File: batch2/create-batch-tables.sql [hunk 13]
```diff
@@ -49,19 +54,34 @@ CREATE TABLE IF NOT EXISTS `jobs` (
...
 CREATE INDEX `jobs_state` ON `jobs` (`state`);
 CREATE INDEX `jobs_instance_name` ON `jobs` (`instance_name`);
 
+CREATE TABLE IF NOT EXISTS `attempts` (
+  `batch_id` BIGINT NOT NULL,
+  `job_id` INT NOT NULL,
+  `attempt_id` INT NOT NULL,
+  `start_time` DOUBLE,
+  `end_time` DOUBLE
+) ENGINE = InnoDB;
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch2/create-batch-tables.sql]: good catch. I thought I added that.

File: batch2/batch/driver/instance_pool.py (hunk 3 above)
Comment (jigold) [reply batch2/batch/driver/instance_pool.py]: If we only care about delete_done events, then I can revert the EntryIterator back to how it was before where it looked at entries in descending order until it reached the mark.

File: batch2/create-batch-tables.sql (hunk 10 above)
Comment (jigold) [reply batch2/create-batch-tables.sql]: Wait, one is OLD and the other is NEW. Is that your objection?

File: batch2/batch/driver/instance_pool.py (hunk 3 above)
Comment (jigold) [reply batch2/batch/driver/instance_pool.py]: > A job will continue to run (and maybe finish) while the preemption/deletion is being processed.

See my comment about left normalizing the time points. I think we'll handle this case correctly the way the code is currently.

File: batch2/batch/batch.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jigold) [reply batch2/batch/batch.py]: oh, I see now

File: batch2/batch/driver/instance_pool.py (hunk 3 above)
Comment (jigold) [reply batch2/batch/driver/instance_pool.py]: I just saw your comment above about billing up until the preemption event. I'm going to keep my code the way it is.

File: batch2/batch/driver/instance.py (hunk 2 above)
Comment (cseed) [reply batch2/batch/driver/instance.py]: I'm afraid I'm confused by this function, I don't see a situation where you would want to set the start to the end, or clear the start (forgetting when it started).  I must be confused about how this function is intended to be used.  Based on the name, why isn't it just setting the end only?

This is called on instance preemption/delete.  If a job started and hour ago, and gets preempted now, the user should get charged for an hour of compute.  But clearing start or setting start = end would appear to lose that information.  What am I missing?

File: batch2/batch/worker.py [hunk 14]
```diff
@@ -212,12 +212,19 @@ def step(self, name, **kwargs):
...
+                        await self.job.mark_started(worker, start_time)
+
                     async with self.step('starting'):
                         await docker_call_retry(self.container.start)
 
                     async with self.step('running'):
                         await docker_call_retry(self.container.wait)
 
+                    if self.name == 'main':
+                        self.job.mark_ended(time.time())
```
Comment (cseed) [inline batch2/batch/worker.py]: This won't run if e.g. wait throws an exception.  We need to capture the time in that case, too.  Also, I think I'd just let the mark complete event set the time instead of sending a separate mark time event.  I think the main runtime timing is actually what we want for the attempt.

File: batch2/batch/worker.py [hunk 15]
```diff
@@ -504,31 +518,53 @@ def id(self):
...
+    async def delete(self, worker):
         log.info(f'deleting {self}')
         self.deleted = True
         for _, c in self.containers.items():
             await c.delete()
 
+        if self.start_time and not self.end_time:
+            await self.mark_cancelled(worker, time.time())
+
+    async def mark_started(self, worker, start_time):
```
Comment (cseed) [inline batch2/batch/worker.py]: If you take my other suggestions, then I think set timing can become set start because it is only used to set the start.

File: batch2/batch/worker.py (hunk 7 above)
Comment (cseed) [reply batch2/batch/worker.py]: So with my other suggestions, I'd be OK making this just retry transients.  The only case we'd lose information is when a job starts, this fails, and then the job is cancelled or preempted.  That seems rare enough to be OK.  And if this fails, that probably means something much more serious is wrong.

File: batch2/create-batch-tables.sql [hunk 16]
```diff
@@ -33,6 +33,8 @@ CREATE TABLE IF NOT EXISTS `batches` (
   `n_failed` INT NOT NULL DEFAULT 0,
   `n_cancelled` INT NOT NULL DEFAULT 0,
   `time_created` TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+  `time_completed` TIMESTAMP,
```
Comment (cseed) [inline batch2/create-batch-tables.sql]: I think all timestamps should be doubles in the database in unix time.  I'm surprised time_created is still like that, I thought I changed it.

File: batch2/create-batch-tables.sql [hunk 17]
```diff
@@ -33,6 +33,8 @@ CREATE TABLE IF NOT EXISTS `batches` (
...
+  `cost` DOUBLE NOT NULL DEFAULT 0,
```
Comment (cseed) [inline batch2/create-batch-tables.sql]: To avoid issues with rounding, I wonder if the cost should be a BIGINT measured in, say, millicore-milliseconds.  That's around 5B core-hrs, or a $50M cost.

File: batch2/batch/driver/instance.py (hunk 2 above)
Comment (jigold) [reply batch2/batch/driver/instance.py]: I can just set the end only. It doesn't affect the ultimate answer (the before update trigger takes care of it), but I agree it's clearer to not set start here at all.

File: batch2/create-batch-tables.sql [hunk 18]
```diff
@@ -94,6 +114,42 @@ CREATE INDEX batch_attributes_key_value ON `batch_attributes` (`key`, `value`(25
...
+CREATE TRIGGER attempts_before_update BEFORE UPDATE ON attempts
+FOR EACH ROW
+BEGIN
+  IF OLD.start_time IS NOT NULL AND (NEW.start_time IS NULL OR NEW.start_time > OLD.start_time) THEN
+    SET NEW.start_time = OLD.start_time;
+  END IF;
+
+  IF OLD.end_time IS NOT NULL AND (NEW.end_time IS NULL OR NEW.end_time > OLD.end_time) THEN
+    SET NEW.end_time = OLD.end_time;
+  END IF;
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch2/create-batch-tables.sql]: My code handles the scenario you state above. I always pick the time point that is the earliest for both start time and end time. So in your example, we start a job running, so start is not Null, but end is Null. Then the job is preempted so we mark the end in the database and add the cost to batches and jobs. Then the end message arrives which I think must be before the preemption time (not accounting for time synchronization errors). The before update trigger says to keep the earlier time point, which will update the New.end_time to the earlier time point. The cost needs to be adjusted for the earlier time point, which I believe is what the second trigger does where it subtracts the old time interval from the new one. So in this case, the cost should be negative in the second update to jobs and batches. 

So, my reasoning is it's easier to just take the earliest time point and then you have (1) always taking precedence and then the earliest of (2) and (3) without explicitly having to do the bookkeeping. I can track the reason for the end time and that might make the code easier to always ensure (1) takes precedence, but I argue that it's not necessary. But I agree the reason would be good to have. I'll have to modify the triggers a bit to make sure I get the right reason on update.

Does this clarify my approach?

File: batch2/batch/worker.py [hunk 19]
```diff
@@ -504,31 +518,53 @@ def id(self):
...
+        assert not self.start_time and not self.end_time
+        self.start_time = start_time
+        asyncio.ensure_future(worker.post_attempt_timing(self))
+
+    async def mark_cancelled(self, worker, end_time):
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch2/batch/worker.py]: This will definitely be cleaner. I'll make this change.

File: batch2/batch/worker.py (hunk 14 above)
Comment (jigold) [reply batch2/batch/worker.py]: ok. so you're proposing that we mark the start_time, but the end_time is start_time + main_container_status['runtime']['duration']? That's fine, but I always worry about getting the times to be in the same timezone/units. I'll just have to double check.

File: batch2/create-batch-tables.sql (hunk 17 above)
Comment (jigold) [reply batch2/create-batch-tables.sql]: I was worried about that, but wanted to get the basic thing right first without having an additional units conversion to worry about it. But I'll add the change now.

File: batch2/create-batch-tables.sql (hunk 18 above)
[the remark this answers is not in the document]
Comment (jigold) [reply batch2/create-batch-tables.sql]: > start is only set by the set_timing call from the worker when the attempt actually starts,

I set start from mark_complete as well. It should be the exact same time as the set_timing call. I could have logic to only set it if it's null, but you get the same answer either way.

[the remark this answers is not in the document]
Comment (jigold) [reply batch2/create-batch-tables.sql]: Just for clarity:
> which will update the New.end_time to the earlier time point

This will keep New.end_time to be New.end_time (no change in the update).

File: batch2/batch/worker.py (hunk 14 above)
Comment (cseed) [reply batch2/batch/worker.py]: There are now start_time and end_time (and duration) in the container status timing structs.  Start and end time are in unix time.  So just take the end time.  All times should be in unix time which is seconds elapsed since the epoch, midnight Jan 1 1970 GMT/UTC, so there is no issue with timezones anywhere.

File: batch2/batch/google_compute.py [hunk 20]
```diff
@@ -17,34 +17,43 @@
...
             try:
                 entry = await anext(self.entries)
-                timestamp = entry.timestamp.timestamp()
-                if not self.latest:
-                    self.latest = timestamp
-                # might miss events with duplicate times
-                # solution is to track resourceId
-                if timestamp <= self.mark:
-                    raise StopAsyncIteration
+                timestamp = entry.timestamp.isoformat() + 'Z'
```
Comment (cseed) [inline batch2/batch/google_compute.py]: You're sure the default order is ascending (older to newer)?

File: batch2/batch/worker.py [hunk 21]
```diff
@@ -212,6 +212,10 @@ def step(self, name, **kwargs):
 
             async with self.step('runtime', state=None):
                 async with worker.cpu_sem(self.cpu_in_mcpu):
+                    if self.name == 'main':
+                        self.job.start_time = time.time()
+                        asyncio.ensure_future(worker.post_attempt_timing(self.job))
```
Comment (cseed) [inline batch2/batch/worker.py]: This should be post_job_started?

File: batch2/batch/worker.py [hunk 22]
```diff
@@ -514,24 +525,37 @@ def id(self):
...
+        cstatuses = {
             name: await c.status() for name, c in self.containers.items()
         }
+        status['container_statuses'] = cstatuses
+
+        main_timings = cstatuses['main']['timing']
+        main_duration = main_timings['starting']['duration'] + main_timings['running']['duration']
+
+        status['start_time'] = self.start_time
+        status['end_time'] = self.start_time + main_duration
```
Comment (cseed) [inline batch2/batch/worker.py]: Main may not have timing, so this isn't safe.  I'm not sure what you're calculating here.  My intention was that the timing 'runtime' was what we wanted to report (it measures the time spent in the cpu semaphore), so you should just be able to grab main['timing']['runtime']['start_time'/'end_time'].

File: batch2/batch/worker.py [hunk 23]
```diff
@@ -212,6 +212,10 @@ def step(self, name, **kwargs):
 
             async with self.step('runtime', state=None):
                 async with worker.cpu_sem(self.cpu_in_mcpu):
+                    if self.name == 'main':
+                        self.job.start_time = time.time()
```
Comment (cseed) [inline batch2/batch/worker.py]: Using self.timing['runtime']['start_time'] will be more consistent.

File: batch2/batch/worker.py [hunk 24]
```diff
@@ -611,7 +635,7 @@ def __init__(self):
         if job is None:
             raise web.HTTPNotFound()
 
-        asyncio.ensure_future(job.delete())
+        job.delete()
```
Comment (cseed) [inline batch2/batch/worker.py]: This will at least need an await.  Curious, why this change?

File: batch2/create-batch-tables.sql [hunk 25]
```diff
@@ -196,6 +267,8 @@ BEGIN
 
     IF actual_n_jobs = expected_n_jobs THEN
       UPDATE batches SET closed = 1 WHERE id = in_batch_id;
+      UPDATE batches SET time_completed = UNIX_TIMESTAMP(NOW(3))
+        WHERE id = in_batch_id AND n_completed = batches.n_jobs;
```
Comment (cseed) [inline batch2/create-batch-tables.sql]: I would prefer the driver pass in the time rather than using getting the time in the database.  This is partially for consistency (of code), that the resolution/precision of the clocks may not be the same (you're not truncating Python time to 3 digits), and partially to rely on as few clocks as possible to minimize clock skew.

File: batch2/create-batch-tables.sql [hunk 26]
```diff
@@ -274,6 +351,7 @@ BEGIN
     UPDATE jobs SET state = 'Ready', instance_name = NULL WHERE batch_id = in_batch_id AND job_id = in_job_id;
     UPDATE ready_cores SET ready_cores_mcpu = ready_cores_mcpu + cur_cores_mcpu;
     UPDATE instances SET free_cores_mcpu = free_cores_mcpu + cur_cores_mcpu WHERE name = cur_job_instance_name;
+    UPDATE attempts SET end_time = new_end_time, reason = new_reason WHERE batch_id = in_batch_id AND job_id = in_job_id;
```
Comment (cseed) [inline batch2/create-batch-tables.sql]: Whoa, this can't be right.  You're setting the end time for all the attempts, some of which might have been preempted a long time ago.

File: batch2/create-batch-tables.sql [hunk 27]
```diff
@@ -307,6 +389,9 @@ BEGIN
     WHERE batch_id = in_batch_id AND job_id = in_job_id;
 
     UPDATE batches SET n_completed = n_completed + 1 WHERE id = in_batch_id;
+    UPDATE batches SET time_completed = UNIX_TIMESTAMP(NOW(3))
+      WHERE id = in_batch_id AND n_completed = batches.n_jobs;
```
Comment (cseed) [inline batch2/create-batch-tables.sql]: Same re: timestamp.

File: batch2/create-batch-tables.sql [hunk 28]
```diff
@@ -94,6 +117,54 @@ CREATE INDEX batch_attributes_key_value ON `batch_attributes` (`key`, `value`(25
...
+
+CREATE TRIGGER attempts_after_update AFTER UPDATE ON attempts
+FOR EACH ROW
+BEGIN
+  DECLARE cores_mcpu INT;
+  DECLARE time_diff DOUBLE;
+  DECLARE cost_per_core_sec DOUBLE;
+  DECLARE new_msec_mcpu BIGINT;
+
+  SET cost_per_core_sec = 0.01 / 3600;
```
Comment (cseed) [inline batch2/create-batch-tables.sql]: As mentioned above, this doesn't account for the cost of the disk.

File: batch2/create-batch-tables.sql [hunk 29]
```diff
@@ -94,6 +117,54 @@ CREATE INDEX batch_attributes_key_value ON `batch_attributes` (`key`, `value`(25
...
+
+  SELECT cores_mcpu INTO cores_mcpu FROM jobs
+  WHERE batch_id = NEW.batch_id AND job_id = NEW.job_id;
+
+  SET time_diff = COALESCE(NEW.end_time - NEW.start_time, 0) -
+                  COALESCE(OLD.end_time - OLD.start_time, 0);
+
+  SET new_msec_mcpu = (time_diff * 1000) * cores_mcpu;
```
Comment (cseed) [inline batch2/create-batch-tables.sql]: Maybe `new_msec_mcpu_diff`?  It is still a diff.

File: batch2/create-batch-tables.sql [hunk 30]
```diff
@@ -32,7 +32,10 @@ CREATE TABLE IF NOT EXISTS `batches` (
   `n_succeeded` INT NOT NULL DEFAULT 0,
   `n_failed` INT NOT NULL DEFAULT 0,
   `n_cancelled` INT NOT NULL DEFAULT 0,
-  `time_created` TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+  `time_created` DOUBLE NOT NULL,
+  `time_completed` DOUBLE,
+  `msec_mcpu` BIGINT NOT NULL DEFAULT 0,
+  `cost` DOUBLE NOT NULL DEFAULT 0
```
Comment (cseed) [inline batch2/create-batch-tables.sql]: `msec_mcpu` is just units, I'd call it `cost_msec_mcpu`.  I wouldn't keep two costs, just convert msec_mcpu when you read it out.

There is another issue, that we're not calculating the cost of the disk.  To start, I think we can keep it as 100G and fold it into the per core-time cost.  That means larger disk  sizes won't be accounted for.  I think that's OK for now.  We can track the disk cost directly when we support requested disk sizes.

File: batch2/create-batch-tables.sql (hunk 28 above)
Comment (cseed) [reply batch2/create-batch-tables.sql]: SSD PD is $0.17/GB-month.  So for 8 cores, 100G, the cost of disk per core hour is:

```
0.170 / 30 / 24 * 12.5
.00295138888888888887
```

So the total instance cost per core-hour is 0.0013.

**Comment by jigold** [comment]
> The end we're taking the max (in our favor). I'm not totally comfortable with that, e.g. a preemption happens after a job is finished, but the preemption event gets recorded first, you'll charge the user up to the preemption (max) even tho the job already finished. That doesn't seem right. So I see two options:

I intended to always take the min (user's favor) for the end time and the start time should never change but if it does I take it to be earlier in our favor. If the trigger isn't doing that then I got it wrong. Basically, the way I understood the trigger is as follows:

```
  IF OLD.end_time IS NOT NULL AND (NEW.end_time IS NULL OR NEW.end_time > OLD.end_time) THEN
    SET NEW.end_time = OLD.end_time;
    SET NEW.reason = OLD.reason;
  END IF;
```

This will update the record to have the new end time unless the old end time is not NULL and either the new end time is null (don't want to overwrite the existing value with a null value) or new end time > old end time (don't want to update the record with an end time that is greater than the existing end time in the database). To not override the values, then we need to set the new end time and reason to the old end time and reason to avoid updating the record.

Does this make sense?  

> Use the reason to update the end time. Completed time should be taken first, then I think whatever is earliest (in the user's favor) between deletion/preemption and cancellation.

I can do this if you think it's clearer. I think the answer will be the same.

> Alternatively, we just take the earliest time (in the user's favor) always which I think can only give up a small amount of compute on deletion/preemption or cancellation (where we see the event, and while it's being processed, the job completes).

This is what I'm doing (or I think I'm doing). See comment above.

File: batch2/batch/worker.py (hunk 24 above)
Comment (jigold) [reply batch2/batch/worker.py]: I didn't mean to change it. I didn't look at the diff carefully enough. Sorry about that.

File: batch2/batch/worker.py (hunk 22 above)
Comment (jigold) [reply batch2/batch/worker.py]: My concern was that acquiring the semaphore occurs after starting the clock for the runtime. I think we can just change the order of those two statements and then I can do the easier thing you suggested.

File: batch2/batch/worker.py (hunk 21 above)
Comment (jigold) [reply batch2/batch/worker.py]: good catch.

File: batch2/batch/google_compute.py (hunk 20 above)
Comment (jigold) [reply batch2/batch/google_compute.py]: I'll add the ordering back to make sure.

File: batch2/create-batch-tables.sql (hunk 28 above)
Comment (jigold) [reply batch2/create-batch-tables.sql]: Shouldn't there be a factor of 100 in there?

0.17 GB/month * 100 GB * (1 month / 30 days) * (1 day / 24 hours) * (1 machine / 8 cores)

Ok. So that's where the 12.5 comes from then.

File: batch2/create-batch-tables.sql (hunk 26 above)
Comment (jigold) [reply batch2/create-batch-tables.sql]: I was assuming the same logic where the minimum end time is always taken. I can do a join with the jobs table, filter to the attempts on that instance and update or go further and look for attempts where the reason is not completed, preempted.

File: batch2/create-batch-tables.sql (hunk 25 above)
Comment (jigold) [reply batch2/create-batch-tables.sql]: ok.

File: batch2/batch/worker.py (hunk 23 above)
Comment (jigold) [reply batch2/batch/worker.py]: My concern was with the cpu_semaphore acquisition not matching the runtime start time. If I switch the order of the cpu_semaphore acquisition with the runtime start, then I'm okay doing this. Otherwise, I worry that for example an 8 core job that never acquires the semaphore, this will be wrong.

**Comment by cseed** [comment]
I'm convinced your original approach was good.  Don't switch on the reason (in fact, I'm OK if you rip it out ... sorry!)  However, it looks like the start time is taking the earlier time (in our favor):

```
  IF OLD.start_time IS NOT NULL AND (NEW.start_time IS NULL OR NEW.start_time > OLD.start_time) THEN
    SET NEW.start_time = OLD.start_time;
  END IF;
```

File: batch2/create-batch-tables.sql (hunk 30 above)
Comment (jigold) [reply batch2/create-batch-tables.sql]: I was concerned that if google's pricing changed some day (or we change our model), then we'd be using the new rate when calculating the cost of older batches. That's why I thought we needed both the amount of resources used and the cost computed from the total number of resources used.

**Comment by jigold** [comment]
> However, it looks like the start time is taking the earlier time (in our favor):

That's correct. I can switch that to take the max. So we always give the user the more favorable timing.

**Comment by jigold** [comment]
The reason might be good to leave in so we can tell what the preemption rate, cancellation rate, etc. are

File: batch2/create-batch-tables.sql (hunk 28 above)
Comment (jigold) [reply batch2/create-batch-tables.sql]: This should be 0.013 (0.01 + 0.003)

**Comment by cseed** [comment]
Yeah, I agree.  Leave it.

File: batch2/create-batch-tables.sql (hunk 28 above)
Comment (cseed) [reply batch2/create-batch-tables.sql]: Yeah, the 12.5 is 100 / 8.

File: batch2/batch/worker.py (hunk 22 above)
Comment (cseed) [reply batch2/batch/worker.py]: Oh, you're totally right.  The timing should be inside the semaphore, that's a total mistake.

File: batch2/batch/worker.py (hunk 23 above)
Comment (cseed) [reply batch2/batch/worker.py]: Yeah, do that, that was a mistake on my end, I didn't mean the runtime to include the wait for resources.

File: batch2/create-batch-tables.sql (hunk 30 above)
Comment (cseed) [reply batch2/create-batch-tables.sql]: I wouldn't worry about that now.  I think this (and full and accurate resource tracking with costs) is going to require a more complicated solution.

File: batch2/create-batch-tables.sql (hunk 26 above)
Comment (jigold) [reply batch2/create-batch-tables.sql]: @cseed I'm not sure what to do here. The more complicated query doesn't buy you anything because the job is already checked to be on that instance_id with the if clause. I think my query is fine.

Comment (cseed) [reply batch2/create-batch-tables.sql]: I think we should store the attempt in the job.  Then you can get the cur_attempt when you query the job, and then reset just that one.

File: batch2/batch/driver/instance.py [hunk 31]
```diff
@@ -105,6 +105,15 @@ def state(self):
         self._state = 'deleted'
         self.instance_pool.adjust_for_add_instance(self)
 
+    async def mark_jobs_ended(self, timestamp, reason):
```
Comment (cseed) [inline batch2/batch/driver/instance.py]: I think you can remove this and fold the behavior into `deactivate_instance`.  This also feels better: we should always be setting the end_time and reason when we're updating it because we know it isn't running anymore (or is about to stop running as the case may be).

File: batch2/batch/worker.py [hunk 32]
```diff
@@ -514,24 +521,39 @@ def id(self):
...
     # }
     async def status(self):
         status = {
             'worker': NAME,
             'batch_id': self.batch_id,
             'job_id': self.job_spec['job_id'],
+            'attempt_id': self.job_spec['attempt_id'],
             'user': self.user,
-            'state': self.state
+            'state': self.state,
```
Comment (cseed) [inline batch2/batch/worker.py]: Just noticed, this isn't necessary.

File: batch2/batch/google_compute.py [hunk 33]
```diff
@@ -17,34 +17,43 @@
...
+        self.mark = timestamp
 
     def __aiter__(self):
         return self
 
     async def __anext__(self):
         while True:
             if not self.entries:
-                self.entries = await self.gservices.list_entries()
+                self.entries = await self.gservices.list_entries(self.mark)
```
Comment (cseed) [inline batch2/batch/google_compute.py]: We should add a call to asyncio.sleep(5) before list_entries.

File: batch2/create-batch-tables.sql [hunk 34]
```diff
@@ -134,14 +194,20 @@ BEGIN
     UPDATE ready_cores
     SET ready_cores_mcpu = ready_cores_mcpu +
       COALESCE(
-        (SELECT SUM(cores_mcpu)
-         FROM jobs
+        (SELECT SUM(jobs.cores_mcpu)
+         FROM attempts
+	 INNER JOIN jobs ON attempts.batch_id = jobs.batch_id AND attempts.job_id = jobs.job_id
```
Comment (cseed) [inline batch2/create-batch-tables.sql]: Spacing looks a bit off here on Github.

File: batch2/batch/driver/instance.py (hunk 31 above)
Comment (jigold) [reply batch2/batch/driver/instance.py]: This is a bit difficult. There's a lot of things that call deactivate_instance. I can try and add it with a default reason of 'unknown' and timestamp = now unless the reason and timestamp have been explicitly passed to the function.

Comment (cseed) [reply batch2/batch/driver/instance.py]: I feel like the cases where we deactivate/unschedule but don't set the end time are actually bugs.

Yes.  But aren't those the real reasons?  I say it should take a reason and an optional timestamp (and use now if it isn't passed).  I realize you might need to pass the reason down a level or two to get the proper reason.  As you say, the metrics on these are going to be interesting.

Comment (jigold) [reply batch2/batch/driver/instance.py]: I'm worried that if we update the timestamp and reason for everything (not just explicitly for preemption/deletion), then it might break my logic where the job could continue running after the instance is marked deactivated (temporary network failure attempting to contact instance) and we won't get the true end time. I'm thinking the defaults should just be None for both reason and timestamp. I can look at where we call deactivate more closely though.

Comment (cseed) [reply batch2/batch/driver/instance.py]: Break your logic where?  In that case, we'll just take the earlier time.  I originally argued for taking the true end time based on the reason, but basically I'm convinced this is super rare and the difference isn't worth the complexity.

Comment (jigold) [reply batch2/batch/driver/instance.py]: I think you're right and this is fine. I added a bunch of reasons, but I don't think we should see any reasons in the attempts table after all the gevents have been processed except 'preempted' and 'deleted'.
